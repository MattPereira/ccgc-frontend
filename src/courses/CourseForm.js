import React, { useState, useContext } from "react";
import UserContext from "../auth/UserContext";
import CcgcApi from "../api/api";
import {
  Card,
  CardBody,
  CardTitle,
  Form,
  Label,
  Input,
  FormGroup,
  Alert,
} from "reactstrap";

/** Form to create (and maybe edit?) a course
 *
 * Displays course form and handles changes to local form state.
 * Submission of form calls the API to save the course and redirects
 * to the courses list page.
 *
 * Redirects to CourseDetails page upon form submission.
 *
 * Routed as /courses/new
 * Routes -> CourseForm -> Alert
 */

///MAKE A COURSE_HANDLE BY SPLITTING NAME BY SPACES AND JOINING WITH '-'

const CourseForm = () => {
  const { currentUser, setCurrentUser } = useContext(UserContext);

  const [formData, setFormData] = useState({
    name: "",
    rating: "",
    slope: "",
    pars: {
      hole1: "",
      hole2: "",
      hole3: "",
      hole4: "",
      hole5: "",
    },
  });

  const [formErrors, setFormErrors] = useState([]);
  const [updateConfirmed, setUpdateConfirmed] = useState(false);

  console.debug(
    "ProfileForm",
    "currentUser=",
    currentUser,
    "formData=",
    formData,
    "formErrors=",
    formErrors,
    "updateConfirmed=",
    updateConfirmed
  );

  //update state of formData onChange of any form input field
  const handleChange = (e) => {
    const { name, value } = e.target;

    setFormData((fData) => ({
      ...fData,
      [name]: value,
    }));
    setFormErrors([]);
  };

  /**on form submission:
   * -attempt save to backend & report any errors
   * -if successful
   *  -clear previous error messages and password
   *  - show update-confirmed alert
   *  - set current user info throughout the site
   */

  const handleSubmit = async (e) => {
    e.preventDefault();

    let profileData = {
      firstName: formData.firstName,
      lastName: formData.lastName,
      email: formData.email,
      password: formData.password,
      bio: formData.bio,
    };

    let updatedUser;

    try {
      updatedUser = await CcgcApi.updateProfile(
        currentUser.username,
        profileData
      );
    } catch (errors) {
      debugger;
      setFormErrors(errors);
      return;
    }

    setFormData((fData) => ({ ...fData, password: "" }));
    setFormErrors([]);
    setUpdateConfirmed(true);

    //trigger reloading of user information throughout the site
    setCurrentUser(updatedUser);
  };

  return (
    <div className="row justify-content-center">
      <div className="col-md-8">
        <Card className="px-5 py-3">
          <CardBody>
            <CardTitle className="display-4 text-center">New Course</CardTitle>

            <Form onSubmit={handleSubmit}>
              <div className="row">
                <div className="">
                  <FormGroup>
                    <Label htmlFor="name">Course Name</Label>
                    <Input
                      className="form-control"
                      id="name"
                      name="name"
                      type="text"
                      onChange={handleChange}
                      value={formData.name}
                      required
                    ></Input>
                  </FormGroup>
                </div>
                <div className="col-lg-6"></div>
              </div>

              <div className="row">
                <div className="col-6">
                  <FormGroup>
                    <Label htmlFor="rating">Rating</Label>
                    <Input
                      className="form-control"
                      id="rating"
                      name="rating"
                      type="number"
                      step="0.1"
                      onChange={handleChange}
                      value={formData.rating}
                      required
                    ></Input>
                  </FormGroup>
                </div>
                <div className="col-6">
                  <FormGroup>
                    <Label htmlFor="slope">Slope</Label>
                    <Input
                      className="form-control"
                      id="slope"
                      name="slope"
                      type="slope"
                      onChange={handleChange}
                      value={formData.slope}
                      required
                    ></Input>
                  </FormGroup>
                </div>
              </div>

              <div className="row">
                <div className="col-2">
                  <label>Hole</label>
                </div>
                <div className="col-5">
                  <Label htmlFor="par1">Par</Label>
                </div>
                <div className="col-5">
                  <Label htmlFor="handicap1">Handicap</Label>
                </div>
              </div>

              <div className="row">
                <div className="col-2">
                  <label>#1</label>
                </div>
                <div className="col-5">
                  <FormGroup>
                    <Input
                      className="form-control"
                      id="par1"
                      name="par1"
                      type="text"
                      onChange={handleChange}
                      value={formData.email}
                      required
                    ></Input>
                  </FormGroup>
                </div>
                <div className="col-5">
                  <FormGroup>
                    <Input
                      className="form-control"
                      id="handicap1"
                      name="handicap1"
                      type="handicap1"
                      onChange={handleChange}
                      value={formData.password}
                      required
                      autoComplete="current-password"
                    ></Input>
                  </FormGroup>
                </div>
              </div>

              <div className="row">
                <div className="col-2">
                  <label className="align-self-center">#2</label>
                </div>
                <div className="col-5">
                  <FormGroup>
                    <Input
                      className="form-control"
                      id="par1"
                      name="par1"
                      type="text"
                      onChange={handleChange}
                      value={formData.email}
                      required
                    ></Input>
                  </FormGroup>
                </div>
                <div className="col-5">
                  <FormGroup>
                    <Input
                      className="form-control"
                      id="handicap1"
                      name="handicap1"
                      type="handicap1"
                      onChange={handleChange}
                      value={formData.password}
                      required
                      autoComplete="current-password"
                    ></Input>
                  </FormGroup>
                </div>
              </div>

              {formErrors.length
                ? formErrors.map((err) => (
                    <Alert key={err} color="danger">
                      {err}
                    </Alert>
                  ))
                : null}

              {updateConfirmed ? (
                <Alert type="success">Profile information updated!</Alert>
              ) : null}

              <div className="row justify-content-end">
                <div className="col-auto">
                  <button className="btn btn-primary btn-block px-4">
                    Submit
                  </button>
                </div>
              </div>
            </Form>
          </CardBody>
        </Card>
      </div>
    </div>
  );
};

export default CourseForm;
