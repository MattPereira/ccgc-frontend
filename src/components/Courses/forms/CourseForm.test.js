import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import CourseForm from "./NewCourseForm";

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <CourseForm />
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <CourseForm />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("displays form inputs and login button", function () {
  render(
    <MemoryRouter initialEntries={["/login"]}>
      <CourseForm />
    </MemoryRouter>
  );

  const nameInput = screen.getByLabelText("Course Name");
  expect(nameInput).toBeInTheDocument();

  const imgUrlInput = screen.getByLabelText("Image Url");
  expect(imgUrlInput).toBeInTheDocument();

  const ratingInput = screen.getByLabelText("Rating");
  expect(ratingInput).toBeInTheDocument();

  const slopeInput = screen.getByLabelText("Slope");
  expect(slopeInput).toBeInTheDocument();

  //only one button in the component
  const submitBtn = screen.getByRole("button");
  expect(submitBtn).toBeInTheDocument();
});
