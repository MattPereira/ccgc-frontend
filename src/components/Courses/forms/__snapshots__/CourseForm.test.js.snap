// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="row justify-content-center"
  >
    <div
      class="col-md-9 col-lg-7"
    >
      <div
        class="card"
      >
        <div
          class="display-4 text-center mb-4 py-2 bg-dark text-white card-title h5"
        >
          Create Course
        </div>
        <div
          class="card-body"
        >
          <form
            class=""
          >
            <div
              class="row"
            >
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                  for="name"
                >
                  Course Name
                </label>
                <input
                  class="form-control"
                  id="name"
                  name="name"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                  for="name"
                >
                  Image Url
                </label>
                <input
                  class="form-control"
                  id="imgUrl"
                  name="imgUrl"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="row justify-content-center"
            >
              <div
                class="col-6"
              >
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="rating"
                  >
                    Rating
                  </label>
                  <input
                    class="form-control"
                    id="rating"
                    name="rating"
                    required=""
                    step="0.1"
                    type="number"
                    value=""
                  />
                </div>
              </div>
              <div
                class="col-6"
              >
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="slope"
                  >
                    Slope
                  </label>
                  <input
                    class="form-control"
                    id="slope"
                    name="slope"
                    required=""
                    type="number"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="row text-center"
            >
              <div
                class="col-2"
              >
                <label
                  class="form-label"
                >
                  Hole
                </label>
              </div>
              <div
                class="col-5"
              >
                <label
                  class="form-label"
                >
                  Par
                </label>
              </div>
              <div
                class="col-5"
              >
                <label
                  class="form-label"
                >
                  Handicap
                </label>
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #1
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par1"
                  name="par1"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap1"
                  name="handicap1"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #2
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par2"
                  name="par2"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap2"
                  name="handicap2"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #3
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par3"
                  name="par3"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap3"
                  name="handicap3"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #4
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par4"
                  name="par4"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap4"
                  name="handicap4"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #5
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par5"
                  name="par5"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap5"
                  name="handicap5"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #6
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par6"
                  name="par6"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap6"
                  name="handicap6"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #7
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par7"
                  name="par7"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap7"
                  name="handicap7"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #8
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par8"
                  name="par8"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap8"
                  name="handicap8"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #9
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par9"
                  name="par9"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap9"
                  name="handicap9"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #10
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par10"
                  name="par10"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap10"
                  name="handicap10"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #11
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par11"
                  name="par11"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap11"
                  name="handicap11"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #12
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par12"
                  name="par12"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap12"
                  name="handicap12"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #13
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par13"
                  name="par13"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap13"
                  name="handicap13"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #14
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par14"
                  name="par14"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap14"
                  name="handicap14"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #15
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par15"
                  name="par15"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap15"
                  name="handicap15"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #16
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par16"
                  name="par16"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap16"
                  name="handicap16"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #17
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par17"
                  name="par17"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap17"
                  name="handicap17"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row align-items-center mb-3"
            >
              <div
                class="col-2 text-center"
              >
                <label
                  class="form-label"
                >
                  #18
                </label>
              </div>
              <div
                class="col-5 align-self-center"
              >
                <input
                  class="form-control"
                  id="par18"
                  name="par18"
                  required=""
                  type="number"
                  value=""
                />
              </div>
              <div
                class="col-5"
              >
                <input
                  class="form-control"
                  id="handicap18"
                  name="handicap18"
                  required=""
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="row justify-content-end"
            >
              <div
                class="col-auto"
              >
                <button
                  class="btn btn-primary btn-block px-4"
                >
                  Submit
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
